### PR TITLE
PLANNER-1607 Make it possible for entities to be Comparable

### PR DESCRIPTION
**Pull request by triceo**
@ge0ffrey This is the last bit of sequential subpillars we talked about.
If possible, please merge on Friday - it actually fixes a bug that we didn't notice until now.

FYI I ran all the examples by hand to make sure that there are no unexpected runtime failures.

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/constraint/ConstraintJustification.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.optaplanner.core.api.score.constraint;
+
+import java.util.Comparator;
+
+public interface ConstraintJustification {
+
+    /**
+     * Used by the GUI to sort the {@link ConstraintMatch} list by {@link ConstraintMatch#getJustificationList()}. Must
+     * be able to meaningfully compare classes of different types, if only by comparing their {@link Class#getName()}.
+     */
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/constraint/ConstraintJustification.java]: Please include @returns in the javadoc (if a javadoc is present, it should be valid)

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/constraint/ConstraintMatch.java [hunk 2]
```diff
@@ -82,7 +83,12 @@ public int compareTo(ConstraintMatch other) {
             return constraintName.compareTo(other.constraintName);
         } else {
             for (int i = 0; i < justificationList.size() && i < other.justificationList.size(); i++) {
-                int comparison = ((Comparable) justificationList.get(i)).compareTo(other.justificationList.get(i));
+                Object left = justificationList.get(i);
+                Object right = other.justificationList.get(i);
+                Comparator comparator = left instanceof ConstraintJustification ?
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/constraint/ConstraintMatch.java]: This confuses me. "left" is a problem fact or a planning entity.
This code reads: "if a problem fact is an instance of ConstraintJustification, it has a method that returns a comperator?

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/domain/solver/TopicConflict.java [hunk 3]
```diff
@@ -87,15 +93,11 @@ public int hashCode() {
 
     @Override
     public int compareTo(TopicConflict other) {
-        return new CompareToBuilder()
-                .append(leftTopic, other.leftTopic)
-                .append(rightTopic, other.rightTopic)
-                .toComparison();
+        return COMPARATOR.compare(this, other);
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/domain/solver/TopicConflict.java]: These kind of changes are a step forward, this should definitely be done.

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/constraint/ConstraintJustification.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package org.optaplanner.core.api.score.constraint;
+
+import java.util.Comparator;
+
+public interface ConstraintJustification {
```
[the remark this answers is not in the document]
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/constraint/ConstraintJustification.java]: It bugs me that problem facts implement an interface which forces them to either contain or build a Comperator.

If they implement this interface, can they not simply implement Comparable?

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/constraint/ConstraintMatch.java (hunk 2 above)
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/constraint/ConstraintMatch.java]: Precisely. By implementing `ConstraintJustification`, you tell this code how to compare the instances without having the instances implement `Comparable<>`.

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/constraint/ConstraintJustification.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/constraint/ConstraintJustification.java]: They can, but then you can not have your own entities `Comparable` using your own `Comparator`. Which is the whole reason we started doing this.

**Comment by triceo** [comment]
Closing as there is going to be a different implementation.

**Pull request closed without merging**